### PR TITLE
HW5

### PR DESCRIPTION
**Pull request by HellHunterMax**
Lets see if i got the correct in 1 try.

### CODE REVIEW

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 1]
```diff
@@ -9,7 +9,40 @@ public static class ArrayOperations
...
+                    if (array[i] > array[k])
+                    {
+                        position++;
+                    }
+                }
+                bool isSameNumberCheck = true;
+
+                while (isSameNumberCheck)
+                {
+                    if (array[i] == 0)
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 11
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter1/ArrayOperations.cs  9
         

Clones added
============
- Src/BootCamp.Chapter1/ArrayOperations.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613&bid=16187418)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 2]
```diff
@@ -88,8 +195,33 @@ public static int[] InsertLast(int[] array, int number)
         /// <returns>A new array with element inserted at a given index. If an array is empty or null, returns new array with number in it.</returns>
         public static int[] InsertAt(int[] array, int number, int index)
         {
-            // ToDo: implement.
-            return array;
+            if (index < 0 || index > array.Length)
+                return array;
+            if (array == null || array.Length == 0)
+                return new int[] { number };
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Remove the array type; it is redundant.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 3]
```diff
@@ -51,8 +116,28 @@ public static int[] RemoveFirst(int[] array)
         /// <returns>A new array with element removed at a given index. If an array is empty or null, returns input array.</returns>
         public static int[] RemoveAt(int[] array, int index)
         {
-            // ToDo: implement.
-            return array;
+            if (array == null || array.Length == 0 || index < 0 || index > array.Length-1)
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 4]
```diff
@@ -9,7 +9,40 @@ public static class ArrayOperations
...
+                }
+                bool isSameNumberCheck = true;
+
+                while (isSameNumberCheck)
+                {
+                    if (array[i] == 0)
+                        isSameNumberCheck = false;
+                    else if (array[i] == newArray[position])
+                        position++;
+                    else
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'else' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 5]
```diff
@@ -51,8 +116,28 @@ public static int[] RemoveFirst(int[] array)
...
+
+            int[] newArray = new int[array.Length - 1];
+            bool isAfterIndex = false;
+
+            for (int i = 0; i < array.Length - 1; i++)
+            {
+                if (i == index)
+                    isAfterIndex = true;
+
+                if (!isAfterIndex)
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Use the '?:' operator here.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 6]
```diff
@@ -88,8 +195,33 @@ public static int[] InsertLast(int[] array, int number)
...
+
+            int[] newArray = new int[array.Length + 1];
+
+            newArray[index] = number;
+
+            bool isAfterIndex = false;
+
+            for (int i = 0; i < array.Length ; i++)
+            {
+                if (i == index)
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 7]
```diff
@@ -9,7 +9,40 @@ public static class ArrayOperations
...
+                        position++;
+                    }
+                }
+                bool isSameNumberCheck = true;
+
+                while (isSameNumberCheck)
+                {
+                    if (array[i] == 0)
+                        isSameNumberCheck = false;
+                    else if (array[i] == newArray[position])
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 8]
```diff
@@ -88,8 +195,33 @@ public static int[] InsertLast(int[] array, int number)
         /// <returns>A new array with element inserted at a given index. If an array is empty or null, returns new array with number in it.</returns>
         public static int[] InsertAt(int[] array, int number, int index)
         {
-            // ToDo: implement.
-            return array;
+            if (index < 0 || index > array.Length)
+                return array;
+            if (array == null || array.Length == 0)
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Change this condition so that it does not always evaluate to 'false'.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 9]
```diff
@@ -63,8 +148,19 @@ public static int[] RemoveAt(int[] array, int index)
         /// <returns>A new array with element added at a given index. If an array is empty or null, returns new array with number in it.</returns>
         public static int[] InsertFirst(int[] array, int number)
         {
-            // ToDo: implement.
-            return array;
+            if (array == null)
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 10]
```diff
@@ -88,8 +195,33 @@ public static int[] InsertLast(int[] array, int number)
         /// <returns>A new array with element inserted at a given index. If an array is empty or null, returns new array with number in it.</returns>
         public static int[] InsertAt(int[] array, int number, int index)
         {
-            // ToDo: implement.
-            return array;
+            if (index < 0 || index > array.Length)
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 11]
```diff
@@ -9,7 +9,40 @@ public static class ArrayOperations
         /// <param name="array">Input array in a random order.</param>
         public static void Sort(int[] array)
         {
-            // ToDo: implement.
+            
+            if (array == null || array.Length == 0)
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 3
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter1/ArrayOperations.cs  9
         

Clones added
============
- Src/BootCamp.Chapter1/ArrayOperations.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613&bid=16187418)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 12]
```diff
@@ -88,8 +219,52 @@ public static int[] InsertLast(int[] array, int number)
...
+            }
+
+            if (index < 0 || index > array.Length)
+            {
+                return array;
+            }
+                
+            if (array.Length == 0)
+            {
+                return new int[] { number };
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Remove the array type; it is redundant.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 13]
```diff
@@ -88,8 +219,52 @@ public static int[] InsertLast(int[] array, int number)
         /// <returns>A new array with element inserted at a given index. If an array is empty or null, returns new array with number in it.</returns>
         public static int[] InsertAt(int[] array, int number, int index)
         {
-            // ToDo: implement.
-            return array;
+            if (array == null)
+            {
+                if (index == 0)
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Use the '?:' operator here.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 14]
```diff
@@ -51,8 +132,32 @@ public static int[] RemoveFirst(int[] array)
...
+            bool isAfterIndex = false;
+
+            for (int i = 0; i < array.Length - 1; i++)
+            {
+                if (i == index)
+                {
+                    isAfterIndex = true;
+                }
+
+                if (!isAfterIndex)
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Use the '?:' operator here.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 15]
```diff
@@ -19,7 +62,23 @@ public static void Sort(int[] array)
         /// <param name="array">Input array in a random order.</param>
         public static void Reverse(int[] array)
         {
-            // ToDo: implement.
+            if (array == null || array.Length == 0)
```
Comment (ellmaster) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: This condition could be a function to save the trouble of writing it over and over again.

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 16]
```diff
@@ -19,7 +62,23 @@ public static void Sort(int[] array)
...
+
+            for (int i = 0; i < array.Length; i++)
+            {
+                newArray[i] = array[array.Length - 1 - i];
+            }
+
+            for (int i = 0; i < newArray.Length; i++)
+            {
+                array[i] = newArray[i];
+            }
```
Comment (ellmaster) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: This looks very inefficient. You can do this better!
You don't even need newArray! :)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 17]
```diff
@@ -9,7 +9,50 @@ public static class ArrayOperations
...
+                    }
+                }
+
+                newArray[position] = array[i];
+
+            }
+            for (int i = 0; i < newArray.Length; i++)
+            {
+                array[i] = newArray[i];
+            }
```
Comment (ellmaster) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: I am not quite convinced that this works for every array. Independent wether this works or not you should provide comments to make your code more understandable! Especially when fancy things are happening.

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 3
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter1/ArrayOperations.cs  9
         

Clones added
============
- Src/BootCamp.Chapter1/ArrayOperations.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613&bid=16187418)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 18]
```diff
@@ -51,8 +126,32 @@ public static int[] RemoveFirst(int[] array)
...
+            bool isAfterIndex = false;
+
+            for (int i = 0; i < array.Length - 1; i++)
+            {
+                if (i == index)
+                {
+                    isAfterIndex = true;
+                }
+
+                if (!isAfterIndex)
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Use the '?:' operator here.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 19]
```diff
@@ -88,8 +213,52 @@ public static int[] InsertLast(int[] array, int number)
...
+            }
+
+            if (index < 0 || index > array.Length)
+            {
+                return array;
+            }
+                
+            if (array.Length == 0)
+            {
+                return new int[] { number };
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Remove the array type; it is redundant.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 20]
```diff
@@ -88,8 +213,52 @@ public static int[] InsertLast(int[] array, int number)
         /// <returns>A new array with element inserted at a given index. If an array is empty or null, returns new array with number in it.</returns>
         public static int[] InsertAt(int[] array, int number, int index)
         {
-            // ToDo: implement.
-            return array;
+            if (array == null)
+            {
+                if (index == 0)
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Use the '?:' operator here.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613)

File: Src/BootCamp.Chapter1/ArrayOperations.cs (hunk 15 above)
Comment (HellHunterMax) [reply Src/BootCamp.Chapter1/ArrayOperations.cs]: personal opinion for me is that this way its more readable and an extra method does not shorten the code, maybe even makes it longer ?
if (CheckForNullORZeroLength(array))
            {
                return;
            }

private static bool CheckForNullORZeroLength(int[] array)
        {
            if (array == null || array.Length == 0)
            {
                return true;
            }
            return false;
        }

File: Src/BootCamp.Chapter1/ArrayOperations.cs (hunk 16 above)
Comment (HellHunterMax) [reply Src/BootCamp.Chapter1/ArrayOperations.cs]: this is now half the length nice one thanks.

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 3
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter1/ArrayOperations.cs  8
         

Clones added
============
- Src/BootCamp.Chapter1/ArrayOperations.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613&bid=16187418)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 21]
```diff
@@ -88,8 +224,52 @@ public static int[] InsertLast(int[] array, int number)
...
+            }
+
+            if (index < 0 || index > array.Length)
+            {
+                return array;
+            }
+                
+            if (array.Length == 0)
+            {
+                return new int[] { number };
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Remove the array type; it is redundant.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 22]
```diff
@@ -51,8 +137,32 @@ public static int[] RemoveFirst(int[] array)
...
+            bool isAfterIndex = false;
+
+            for (int i = 0; i < array.Length - 1; i++)
+            {
+                if (i == index)
+                {
+                    isAfterIndex = true;
+                }
+
+                if (!isAfterIndex)
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Use the '?:' operator here.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 23]
```diff
@@ -88,8 +224,52 @@ public static int[] InsertLast(int[] array, int number)
         /// <returns>A new array with element inserted at a given index. If an array is empty or null, returns new array with number in it.</returns>
         public static int[] InsertAt(int[] array, int number, int index)
         {
-            // ToDo: implement.
-            return array;
+            if (array == null)
+            {
+                if (index == 0)
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Use the '?:' operator here.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613)

**Comment by Almantask** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 3
           

Complexity increasing per file
==============================
- Src/BootCamp.Chapter1/ArrayOperations.cs  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613&bid=16187418)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 24]
```diff
@@ -88,8 +184,52 @@ public static int[] InsertLast(int[] array, int number)
...
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Use the '?:' operator here.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 25]
```diff
@@ -88,8 +184,52 @@ public static int[] InsertLast(int[] array, int number)
...
+            }
+
+            if (index < 0 || index > array.Length)
+            {
+                return array;
+            }
+                
+            if (array.Length == 0)
+            {
+                return new int[] { number };
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Remove the array type; it is redundant.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613)

File: Src/BootCamp.Chapter1/ArrayOperations.cs [hunk 26]
```diff
@@ -51,8 +121,32 @@ public static int[] RemoveFirst(int[] array)
...
+            bool isAfterIndex = false;
+
+            for (int i = 0; i < array.Length - 1; i++)
+            {
+                if (i == index)
+                {
+                    isAfterIndex = true;
+                }
+
+                if (!isAfterIndex)
```
Comment (Almantask) [inline Src/BootCamp.Chapter1/ArrayOperations.cs]: Codacy Issue found: [Use the '?:' operator here.](https://app.codacy.com/manual/Almantask/CSharp-From-Zero-To-Hero/pullRequest?prid=4886613)

File: Src/BootCamp.Chapter1/ArrayOperations.cs (hunk 17 above)
Comment (HellHunterMax) [reply Src/BootCamp.Chapter1/ArrayOperations.cs]: added comments. I did more tests, seems to work every time. Don't know where it might not work.

File: Src/BootCamp.Chapter1/ArrayOperations.cs (hunk 15 above)
Comment (Almantask) [reply Src/BootCamp.Chapter1/ArrayOperations.cs]: Edge case, it is good enough.
